Add test ruckig joint limiter

gwalck opened the pull request:
This is the first draft of adding tests to ruckig joint limiter.
The test do not all pass and I spend half a day trying to understand what to expect from ruckig to make suitable tests. The remaining test that do not pass might be a bug in ruckig_joint_limiter OR rucking. Not sure.

gwalck committed:
Fixed ruckig linking

gwalck committed:
Fixed max_jerk initialization

gwalck committed:
Added protection from invalid desired_positions

gwalck committed:
Removed misleading comment

gwalck committed:
Fixed return value to allow working results

gwalck committed:
Added tests (some fail for unknown reason)

gwalck committed:
Added limitation for max vel and max acc

destogl commented on joint_limits_enforcement_plugins/src/ruckig_joint_limiter.cpp:
```diff
@@ -170,7 +212,7 @@ bool RuckigJointLimiter<joint_limits::JointLimits>::on_enforce(
   }
 
-  return result == ruckig::Result::Finished;
+  return (result == ruckig::Result::Working || result == ruckig::Result::Finished);
```
is it Ok if it is still working and we return? Is this expected?

gwalck replied:
I copied what was in https://github.com/SchillingRobotics/ros2_controllers-stogl/blob/eb1cd2b8693e947fc79335dac5adc9006b895b79/joint_trajectory_controller/src/trajectory.cpp#L632

gwalck replied:
Additionally I can say that `update` in Ruckig lib, will most probably return `ruckig::Result::Working` since one needs multiple `update` calls  to get the `ruckig::Result::Finished` as is done in the while loops in the ruckig tests https://github.com/pantor/ruckig/blob/master/test/test-target.cpp#L141
The `ruckig joint limiter` is not using Ruckig to generate a trajectories, so I suppose a single `update` already applied the correct _profiles_ (not sure what multiple calls do to be honest)

destogl commented on joint_limits_enforcement_plugins/test/test_ruckig_joint_limiter.cpp:
```diff
@@ -12,33 +12,276 @@
+  }
+}
+
+/* no pos enforcing available in ruckig
```
we can leave here and expect ruckig to fail with those values. Then we know that we handle those situations properly.

gwalck replied:
> we can leave here

I suppose you mean we can activate this test and expect the limiter to NOT limit.

destogl commented on joint_limits_enforcement_plugins/test/test_ruckig_joint_limiter.cpp:
```diff
@@ -12,33 +12,276 @@
+    );
+  }
+}
+/* ruckig does not use max_dec anyway
```
can we set this? I think it is not possible, then OK; we can remove this and add a warning in the code if someone defines max_deceleration

gwalck replied:
min_acceleration is part of Ruckig pro apparently (at least it is an std::optional and I had an error which told the variable does not exist) https://github.com/pantor/ruckig/blob/master/include/ruckig/input_parameter.hpp#L91C49-L91C65

gwalck committed:
Added a warning for deceleration limits

gwalck committed:
Activated pos test which expect no limits

gwalck committed:
Removed deceleration test (not supported)

gwalck committed:
Fixed negative velocity test

Split test for positive and negative vel
to have same initial conditions

gwalck committed:
Removed brake test which cannot be triggered

gwalck committed:
Fixed acceleration test

Requires velocity to be set

gwalck committed:
Avoid failures due to double precision errors

gwalck committed:
Fixed enforce in case ruckig fails

desired_joint_states should not be changed if ruckig failed

gwalck committed:
Added and fixed tests

when current_state differs from configured state
enforce current_state is required to get
correct computations

gwalck commented:
Tests were fixed or modified to accommodate to what Ruckig can do. 

Also current_state must be set in configure in case it is different than the current_state at first configure. This was overseen because differs from intuitive behavior (current_state is passed to enforce) and led to failing test. Now fixed. 

Maybe one should add a check within enforce to see if the internal rucking current state drifts too much away from current_state that is passed to enforce and at least warn ?
